### DISCUSSION TITLE
problem to define a parameter

### DISCUSSION THREAD

**Discussion by sSajjad90** [idaholab/moose#17858] [Q&A Modules: General] [answered]
Hi,
I am working on a numerical code to develop the material solution and have to define many variables that read and write data. I have got a problem defining this kind of parameter that did not provide the proper results. Would you please let me what is wrong in defining the code? 

A simple part of code and the results are in the following:
for example, a simple parameter like "_count"  is defined  in a part of the code which must be started from "0" and growth with "1" for each iteration, but it provides the wrong data
I defined a public class of material in the *.h file

#pragma once

#include "Material.h"

class last : public Material
{
public:
  static InputParameters validParams();

  last(const InputParameters & parameters);

protected:
  virtual void computeQpProperties() override;
  virtual void initQpStatefulProperties() override;

  MaterialProperty<Real> & _C_star;

  Real _count;

  

the relevant parts of *.C file :

    _count++;
    _C_star[_qp] = _count;

Result:
----------------+----------------+
| time           | C_star         |
+----------------+----------------

|   5.000000e-01 |    2.305000e+02 |   
|   1.000000e+00 |   4.905000e+02 | 
|   1.500000e+00 |   7.505000e+02 |  
|   2.000000e+00 |   1.010500e+03 | 
|   2.500000e+00 |   1.270500e+03 | 
|   3.000000e+00 |   1.530500e+03 |   
|   3.500000e+00 |   1.790500e+03 |   
|   4.000000e+00 |   2.050500e+03 |  
|   4.500000e+00 |   2.310500e+03 |   
|   5.000000e+00 |   2.570500e+03 | 
|   5.500000e+00 |   2.830500e+03 |  
|   6.000000e+00 |   3.090500e+03 |   
|   6.500000e+00 |   3.350500e+03 |  
|   7.000000e+00 |   3.610500e+03 | 
|   7.500000e+00 |   3.870500e+03 |  

I suppose I have to add a "GenericConstantMaterial" class to define these types of variables. Could you let me know how can I do it in *.h and *.C files?

Thanks in advance,
Sajjad

**Comment by GiudGiud** [comment, marked as answer]
Hi Sajjad

I think one problem is that those lines are executed more than once per time step. The material properties are evaluated many times, at initialization, at every non-linear iteration, for residuals, for jacobians etc.
You could add a check based on the current time step and the last time step _count was incremented. This would limit the number of times _count is incremented

Best,
Guillaume

**Reply by sSajjad90** [reply]
Hi,
Many thanks,
Would you please let me know how can add the check based on the current time step and the last time step _count?

I used the "ElementAverageValue" posts processor that may cause the problem. I checked the different types of executes but they did not work properly.

Any suggestion?

**Reply by GiudGiud** [reply]
You could add a parameter like _current_timestep and have an if condition :
```
if (_t != _current_timestep)
{
  // do the increment
  
  _current_timestep = _t;
}
```
Is _C_star going to be constant across the domain?
If not you ll need _current_timestep to be a field not just a single value

**Reply by sSajjad90** [reply]
Dear Guillaume,
would you please add the command you already told me to the below code? I have got stock to control the number of iteration. In the following, there is a very simple code test just to evaluate _count++ and you could see *.C, *.h and the *.i files.


### test.h file:
```
#pragma once

#include "Material.h"

/**
 *
 */
class test : public Material
{
public:
  static InputParameters validParams();

  test(const InputParameters & parameters);

protected:
  // virtual void initQpStatefulProperties() override;
  void computeQpProperties();


  int _count;
  MaterialProperty<Real> & _C_star;

};

```

### test.C file:
```
#include "test.h"

registerMooseObject("precipitationApp", test);

InputParameters
test::validParams()
{
  InputParameters params = Material::validParams();
  params.addRequiredParam<std::string>("f_name_Cstar", "name of the ac function.");

  return params;
}

test::test(const InputParameters & parameters)
  : Material(parameters),
  _C_star(declareProperty<Real>(getParam<std::string>("f_name_Cstar"))),
  _count(0)
{
}
// void
// test::initQpStatefulProperties()
// {
//
// }
void
test::computeQpProperties()
{
  _count++;
  _C_star[_qp]= _count;
}

```


### test.i file:

```
[Mesh]
  type = GeneratedMesh
  dim = 1
  xmin = 0
  xmax = 1
  nx = 10
[]
[Problem]
  kernel_coverage_check = false
  solve = false
[]

[AuxVariables]
  [./prop1]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]
[AuxKernels]
  [./prop1_output]
    type = MaterialRealAux
    variable = prop1
    property = C_star
    execute_on = 'TIMESTEP_END'
  [../]
[]
[Materials]
  [./MixedModeFiniteSystem]
    type = test
    f_name_Cstar = 'C_star'
    f_name_2Cstar = 'C_2star'
  [../]
[]
[Postprocessors]
  [./C_star]
    type = ElementAverageValue
    variable = prop1
    execute_on = 'TIMESTEP_END'
  [../]
  [./C_star2]
    type = ElementAverageMaterialProperty
    mat_prop = 'C_star'
    execute_on = 'TIMESTEP_END'
  [../]


[]
[Executioner]
  type = Transient
  solve_type = PJFNK
  start_time = 0
  end_time = 5
  dt = 1

[]

[Outputs]
exodus = true
csv = true
[]

```


[test.zip](https://github.com/idaholab/moose/files/6505365/test.zip)

**Reply by GiudGiud** [reply]
Closing this. Hope you managed to implement the pseudo code yourself
